File: solutions/csharp/forth/1/Forth.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;

public class Forth
{
	private static readonly Regex NewDefinitionPattern = new("^:(.*);$", RegexOptions.Compiled);

	private static readonly Dictionary<string, Func<Context, Context>> _defaultHandlerByToken
		= new(StringComparer.OrdinalIgnoreCase){
			{ "+", MathHelpers.Add },
			{ "-", MathHelpers.Subtract },
			{ "*", MathHelpers.Multiply },
			{ "/", MathHelpers.Divide },
			{ "dup", StackHelpers.DuplicateLast },
			{ "drop", StackHelpers.DropLast },
			{ "swap", StackHelpers.SwapLast },
			{ "over", StackHelpers.PenultimateValueCopy },
		};

	public static string Evaluate(string[] instructions) =>
		instructions.Aggregate(new Context(), Evaluate)
			.ToString();

	private static Context Evaluate(Context context, string instruction)
	{
		Match instructionMatch = NewDefinitionPattern.Match(instruction);
		return instructionMatch.Success
			? context.DefineHandler(instructionMatch.Groups[1])
			: FollowInstruction(context, instruction);
	}

	private static Context FollowInstruction(Context context, string instruction)
	{
		List<string> tokens = instruction.Split(' ')
					.ToList();

		while (tokens.Any())
		{
			string nextOperation = ShiftTokensToStack(context, tokens);
			context = Do(context, nextOperation);
		}

		return context;
	}

	private static string ShiftTokensToStack(Context context, List<string> tokens)
	{
		string nextOperation = null;
		do
		{
			string token = tokens.Shift();
			if (int.TryParse(token, out int value))
			{
				context.Stack.Add(value);
			}
			else
			{
				nextOperation = token;
			}
		}
		while (tokens.Any() && string.IsNullOrEmpty(nextOperation));
		return nextOperation;
	}

	private static Context Do(Context context, string operation) =>
		string.IsNullOrEmpty(operation)
			? context
			: context.InstructionsByWord.TryGetValue(operation, out string instruction)
				? FollowInstruction(context, instruction)
				: _defaultHandlerByToken.TryGetValue(operation, out Func<Context, Context> function)
					? function(context)
					: throw new InvalidOperationException($"Unknown operation: `{operation}`.");
}

public class Context
{
	public Dictionary<string, string> InstructionsByWord { get; init; } = new(StringComparer.OrdinalIgnoreCase);

	public List<int> Stack { get; init; } = new();

	public override string ToString() =>
		string.Join(' ', Stack);

	public Context DefineHandler(Group group)
	{
		List<string> definition = group.Value.Split(' ')
			.Where(x => !string.IsNullOrEmpty(x))
			.ToList()
			.RequireAtLeastTwoValues();

		string word = definition.Shift();
		if (int.TryParse(word, out int _))
		{
			throw new InvalidOperationException("Redefining numbers not allowed.");
		}

		InstructionsByWord[word] = CreateInstruction(definition);
		return this;
	}

	private string CreateInstruction(List<string> definition)
	{
		StringBuilder builder = new();
		foreach (string word in definition)
		{
			_ = InstructionsByWord.TryGetValue(word, out string instruction)
				? builder.Append($"{instruction} ")
				: builder.Append($"{word} ");
		}

		return builder.ToString()
			.Trim();
	}

	public Context WithStack(Action<List<int>> function)
	{
		function(Stack);
		return this;
	}
}

public class MathHelpers
{
	public static Context Add(Context context) =>
		DoMath<int>(context, (x, y) => x + y);

	public static Context Subtract(Context context) =>
		DoMath<int>(context, (x, y) => x - y);

	public static Context Multiply(Context context) =>
		DoMath<int>(context, (x, y) => x * y);

	public static Context Divide(Context context) =>
		DoMath<int>(context, (x, y) => x / y);

	private static Context DoMath<T>(Context context, Func<int, int, int> function) =>
		context.WithStack(stack =>
			{
				int x = stack.RequireExactlyTwoValues()
					.Shift();
				int y = stack.Shift();
				int result = function(x, y);
				stack.Add(result);
			});
}

public class StackHelpers
{
	public static Context DuplicateLast(Context context) =>
		context.WithStack(stack =>
		{
			int last = stack.RequireAtLeastOneValue()
				.LastOrDefault();
			stack.Add(last);
		});

	public static Context DropLast(Context context) =>
		context.WithStack(stack =>
			_ = stack.RequireAtLeastOneValue()
				.Pop()
		);

	public static Context SwapLast(Context context) =>
		context.WithStack(stack =>
		{
			int lastValue = stack.RequireAtLeastTwoValues()
				.Pop();
			int penultimateValue = stack.Pop();
			stack.Add(lastValue);
			stack.Add(penultimateValue);
		});

	public static Context PenultimateValueCopy(Context context) =>
		context.WithStack(stack =>
		{
			int penultimateValue = stack.RequireAtLeastTwoValues()[^2];
			stack.Add(penultimateValue);
		});
}

public static class ListExtensions
{
	public static List<T> RequireAtLeastOneValue<T>(this List<T> values) =>
		values.RequiresCount(count => count < 1, "Operation requires at least one value.");

	public static List<T> RequireAtLeastTwoValues<T>(this List<T> values) =>
		values.RequiresCount(count => count < 2, "Operation requires at least two values.");

	public static List<T> RequireExactlyTwoValues<T>(this List<T> values) =>
		values.RequiresCount(count => count != 2, "Operation requires exactly two values.");

	private static List<T> RequiresCount<T>(this List<T> values, Func<int, bool> validation, string errorMessage) =>
		validation(values.Count)
			? throw new InvalidOperationException(errorMessage)
			: values;

	public static T Shift<T>(this List<T> values) =>
		TakeOne(values, x => x.FirstOrDefault());

	public static T Pop<T>(this List<T> values) =>
		TakeOne(values, x => x.LastOrDefault());

	private static T TakeOne<T>(this List<T> values, Func<List<T>, T> function)
	{
		T item = function(values);
		if (item != null)
		{
			_ = values.Remove(item);
		}

		return item;
	}
}

File: solutions/ocaml/forth/4/forth.ml
let new_definition_pattern (string : string) : string option =
  let is_long_enough = String.length string >= 2 in
  let starts_with_colon = string.[0] = ':' in
  let ends_with_semicolon = string.[String.length string - 1] = ';' in
  let is_valid_definition = is_long_enough && starts_with_colon && ends_with_semicolon in
  if is_valid_definition
    then Some (String.sub string 1 (String.length string - 2))
    else None


let default_handler_by_token : (string, Context.t -> Context.t) Hashtbl.t =
  let tbl = Hashtbl.create 16 in
  let add key fn =
    Hashtbl.replace tbl (String.lowercase_ascii key) fn
  in
  add "+" Math_helpers.add;
  add "-" Math_helpers.subtract;
  add "*" Math_helpers.multiply;
  add "/" Math_helpers.divide;
  add "dup" Stack_helpers.duplicate_last;
  add "drop" Stack_helpers.drop_last;
  add "swap" Stack_helpers.swap_last;
  add "over" Stack_helpers.penultimate_value_copy;
  tbl

let rec do_while (body : unit -> unit) (cond : unit -> bool) : unit =
  body ();
  if cond ()
    then do_while body cond

let shift_tokens_to_stack (context : Context.t) (tokens : string list ref) : string =
  let next_operation = ref "" in
  let body () =
    let token = List_helpers.shift tokens in
    match int_of_string_opt token with
    | Some value ->
        context.stack := !(context.stack) @ [value]
    | None ->
        next_operation := token
  in
  let cond () = !tokens <> [] && !next_operation = "" in
  do_while body cond;
  !next_operation

let rec follow_instruction (context : Context.t) (instruction : string) : Context.t =
  let tokens : string list ref =
    instruction
    |> String.split_on_char ' '
    |> List.filter (fun x -> x <> "")
    |> ref
  in

  let rec loop () =
    if !tokens <> [] then begin
      let next_operation = shift_tokens_to_stack context tokens in
      ignore (do_instruction context next_operation);
      loop ()
    end
  in

  loop ();
  context

and do_instruction (context : Context.t) (operation : string) : Context.t =
  if operation = "" then
    context
  else
    match Context.get_instruction context operation with
    | Some instruction -> follow_instruction context instruction
    | None ->
        match Hashtbl.find_opt default_handler_by_token (String.lowercase_ascii operation) with
        | Some f -> f context
        | None ->
            failwith (Printf.sprintf "Unknown operation: `%s`." operation)

let evaluate_instruction (context : Context.t) (instruction : string) : Context.t =
  match new_definition_pattern instruction with
  | Some definition -> Context.define_handler context { Definition_string.value = definition }
  | None -> follow_instruction context instruction

let evaluate (instructions : string list) : int list option =
  try
    let final_context = List.fold_left evaluate_instruction (Context.create ()) instructions in
    Some !(final_context.Context.stack)
  with _ ->
    None
